python/lsst/daf/butler/core/butlerConfig.py

```diff
@@ -0,0 +1,126 @@
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Configuration classes specific to the Butler
+"""
+
+import os
+
+from .config import Config
+from .datastore import DatastoreConfig
+from .schema import SchemaConfig
+from .registry import RegistryConfig
+from .storageClass import StorageClassConfig
+from .utils import doImport
+
+__all__ = ("ButlerConfig",)
+
+
+class ButlerConfig(Config):
+    """Contains the configuration for a `Butler`
+
+    The configuration is read and merged with default configurations for
+    the particular classes. The defaults are read from
+    ``$DAF_BUTLER_DIR/config`` and ``$DAF_BULTER_CONFIG_PATH``. The defaults
+    are constructed by reading first the global defaults, and then adding
+    in overrides from each entry in the colon-separated
+    ``$DAF_BUTLER_CONFIG_PATH`` in reverse order such that the entries ahead
+    in the list take priority. The registry and datastore configurations
+    are read using the names specified by the appropriate classes defined
+    in the supplied butler configuration.
+
+    The externally supplied butler configuration must include definitions
+    for ``registry.cls`` and ``datastore.cls`` to enable the correct defaults
+    to be located.
+
+    Parameters
+    ----------
+    other : `str`
+        Path to butler configuration YAML file.
+    """
+
+    def __init__(self, other):
+
+        # Create an empty config for us to populate
+        super().__init__()
+
+        # Read the supplied config so that we can work out which other
+        # defaults to use.
+        butlerConfig = Config(other)
+
+        # All the configs that can be associated with default files
+        configComponents = (SchemaConfig, StorageClassConfig, DatastoreConfig, RegistryConfig)
+
+        # This is a list of all the config files to search
+        defaultsFiles = [(c, c.defaultConfigFile) for c in configComponents]
+
+        # Components that derive their configurations from implementation
+        # classes rather than configuration classes
+        specializedConfigs = (DatastoreConfig, RegistryConfig)
+
+        # Look for class specific default files to be added to the file list
+        # These class definitions must come from the suppliedbutler config and
+        # are not defaulted.
+        for componentConfig in specializedConfigs:
```

Comment:
Why not simply loop over all components and have those subclasses determine if they are special?

Reply:
E.g. by having a `component` member.

Reply:
Are you saying I should look through all the top level keys in the supplied butler config looking for `.cls` members, then ask the imported classes for their config files and associated config classes? (the latter they don't know but they could).